### PR TITLE
Add polish to rcmgr implementation

### PR DESCRIPTION
**Pull request by vyzo**
Changes stemming from raul's review and requests:
- Renames Transaction to Span
- Introduces Protocol Peer Limits
- Don't remove protocol from the constraint set when setting the service
- Rename constraints to edges
- Add logging, remove panics
- Add total stream and conn limit
- Extension APIs for introspection
- Tracing
- Config parser for easy limiter construction using json config
- Refactor defaults to be easily accessible and stop repeating the memory constants.

### CODE REVIEW

File: extapi.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func (s *resourceScope) SetLimit(limit Limit) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.rc.limit = limit
+}
+
+func (r *resourceManager) ListServices() []string {
```
Comment (marten-seemann) [inline extapi.go:50]: Should we sort the result in all `List*` functions? Map iterations are (pseudo-) random.

File: extapi.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	for p := range r.peer {
+		result = append(result, p)
+	}
+
+	return result
+}
+
+func (r *resourceManager) Stat() (result ResourceManagerStat) {
+	r.mx.Lock()
+	svcs := make([]*serviceScope, 0, len(r.svc))
```
Comment (marten-seemann) [inline extapi.go:88]: Same here.

**Review by marten-seemann — APPROVED** [review]

File: extapi.go (hunk 1 above)
Comment (vyzo) [reply extapi.go:50]: we could i guess, why not.

File: extapi.go (hunk 2 above)
Comment (vyzo) [reply extapi.go:88]: no point in sorting here, we put them in a map for the result anyway.

File: extapi.go (hunk 1 above)
Comment (vyzo) [reply extapi.go:54]: done.

**Pull request merged**